# SQL in one lesson
![exploits_of_a_mom.png](exploits_of_a_mom.png)

A lot of the data that we interact with today is stored in databases. For example:

- Student records, including grades, at a school
- Posts and friends in your favorite social network
- News stories on a newspaper's website
- Your contacts list on your mobile phone
- All images that make up Google Maps

All these bits of information are stored in various kinds of databases. Some of these are stored in relational databases that are available as open source tools like Postgresql, MySQL and SQLite.

Others are stored in proprietary systems like Google's [BigTable](https://en.wikipedia.org/wiki/Bigtable) or Facebook's [Haystack Object Store](https://code.fb.com/core-data/needle-in-a-haystack-efficient-storage-of-billions-of-photos/).

Whatever the database might, there needs to be a way to extract data from it and a lot of these systems have agreed on a shared language for accessing data.

This language is called SQL (Structured Query Language, pronounced like "sequel").

To try out SQL, we're going to be using a data set about sales of video games.

Let's start by connecting to the database and taking a peek at the table called `vgsale`.

In [1]:
%load_ext sql

In [2]:
%%sql postgresql://apicard:gin2ger@localhost/video_games
SELECT * FROM vgsale LIMIT 10


10 rows affected.


[(1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74),
 (2.0, 'Super Mario Bros.', 'NES', 1985, 'Platform', 'Nintendo', 29.08, 3.58, 6.81, 0.77, 40.24),
 (3.0, 'Mario Kart Wii', 'Wii', 2008, 'Racing', 'Nintendo', 15.85, 12.88, 3.79, 3.31, 35.82),
 (4.0, 'Wii Sports Resort', 'Wii', 2009, 'Sports', 'Nintendo', 15.75, 11.01, 3.28, 2.96, 33.0),
 (5.0, 'Pokemon Red/Pokemon Blue', 'GB', 1996, 'Role-Playing', 'Nintendo', 11.27, 8.89, 10.22, 1.0, 31.37),
 (6.0, 'Tetris', 'GB', 1989, 'Puzzle', 'Nintendo', 23.2, 2.26, 4.22, 0.58, 30.26),
 (7.0, 'New Super Mario Bros.', 'DS', 2006, 'Platform', 'Nintendo', 11.38, 9.23, 6.5, 2.9, 30.01),
 (8.0, 'Wii Play', 'Wii', 2006, 'Misc', 'Nintendo', 14.03, 9.2, 2.93, 2.85, 29.02),
 (9.0, 'New Super Mario Bros. Wii', 'Wii', 2009, 'Platform', 'Nintendo', 14.59, 7.06, 4.7, 2.26, 28.62),
 (10.0, 'Duck Hunt', 'NES', 1984, 'Shooter', 'Nintendo', 26.93, 0.63, 0.28, 0.47, 28.31)]

The first line of that code block is just a magic invocation that connects us to the database we are going to use.

The second line introduces SQL syntax for the first time. To help you understand the SQL commands we are using,
the SQL syntax words are listed in CAPITAL letters, the lowercase words are the names of tables or columns.

The above statement translates to: grab (SELECT) all the values (*) in the table called vgsale (FROM vgsale) but only
show me the first ten (LIMIT 10).

The table is composed of several columns:

- rank: The sales rank of that game
- name: The name of the game
- platform: The gaming system that it was published for
- year: The year that the game was published
- genre: The genre of the game
- publisher: The company that published the game
- na_sales: Millions of copies sold in North America
- eu_sales: Millions of copies sold in Europe
- jp_sales: Millions of copies sold in Japan
- other_sales: Millions of copies sold in the rest of the world
- global_sales: Millions of copies sold in total


We don't always want to read all the columns in a table. For example, if we just want the name, year and global sales numbers we could select:

In [3]:
%%sql
SELECT name, year, global_sales
FROM vgsale
LIMIT 10

 * postgresql://apicard:***@localhost/video_games
10 rows affected.


[('Wii Sports', 2006, 82.74),
 ('Super Mario Bros.', 1985, 40.24),
 ('Mario Kart Wii', 2008, 35.82),
 ('Wii Sports Resort', 2009, 33.0),
 ('Pokemon Red/Pokemon Blue', 1996, 31.37),
 ('Tetris', 1989, 30.26),
 ('New Super Mario Bros.', 2006, 30.01),
 ('Wii Play', 2006, 29.02),
 ('New Super Mario Bros. Wii', 2009, 28.62),
 ('Duck Hunt', 1984, 28.31)]

SQL doesn't care about line breaks so we can spread a SQL query over multiple lines just to make it easier to read.

## Filtering

We've seen how to look only at certain columns of the table but it is often useful to only look at certain rows in a table. For example, we could want to look only at the top selling games that have been released since you've been playing video games. Let's say you've been playing since 2010:

In [4]:
%%sql

SELECT rank, name, year, publisher, platform
FROM vgsale
WHERE year >= 2010 
LIMIT 20;


 * postgresql://apicard:***@localhost/video_games
20 rows affected.


[(16.0, 'Kinect Adventures!', 2010, 'Microsoft Game Studios', 'X360'),
 (17.0, 'Grand Theft Auto V', 2013, 'Take-Two Interactive', 'PS3'),
 (24.0, 'Grand Theft Auto V', 2013, 'Take-Two Interactive', 'X360'),
 (27.0, 'Pokemon Black/Pokemon White', 2010, 'Nintendo', 'DS'),
 (30.0, 'Call of Duty: Modern Warfare 3', 2011, 'Activision', 'X360'),
 (32.0, 'Call of Duty: Black Ops', 2010, 'Activision', 'X360'),
 (33.0, 'Pokemon X/Pokemon Y', 2013, 'Nintendo', '3DS'),
 (34.0, 'Call of Duty: Black Ops 3', 2015, 'Activision', 'PS4'),
 (35.0, 'Call of Duty: Black Ops II', 2012, 'Activision', 'PS3'),
 (36.0, 'Call of Duty: Black Ops II', 2012, 'Activision', 'X360'),
 (38.0, 'Call of Duty: Modern Warfare 3', 2011, 'Activision', 'PS3'),
 (41.0, 'Call of Duty: Black Ops', 2010, 'Activision', 'PS3'),
 (43.0, 'Mario Kart 7', 2011, 'Nintendo', '3DS'),
 (45.0, 'Grand Theft Auto V', 2014, 'Take-Two Interactive', 'PS4'),
 (50.0, 'Pokemon Omega Ruby/Pokemon Alpha Sapphire', 2014, 'Nintendo', '3DS'),
 (54.0, 'Super Mario 3D Land', 2011, 'Nintendo', '3DS'),
 (55.0, 'Gran Turismo 5', 2010, 'Sony Computer Entertainment', 'PS3'),
 (61.0, 'Just Dance 3', 2011, 'Ubisoft', 'Wii'),
 (62.0, 'Call of Duty: Ghosts', 2013, 'Activision', 'X360'),
 (63.0, 'Halo: Reach', 2010, 'Microsoft Game Studios', 'X360')]

It's also possible to filter by multiple criteria. For example to look at only XBox 360 games released since 2010:

In [5]:
%%sql
SELECT rank, name, year, publisher, platform
FROM vgsale
WHERE year >= 2010
AND platform = 'X360'
LIMIT 20

 * postgresql://apicard:***@localhost/video_games
20 rows affected.


[(16.0, 'Kinect Adventures!', 2010, 'Microsoft Game Studios', 'X360'),
 (24.0, 'Grand Theft Auto V', 2013, 'Take-Two Interactive', 'X360'),
 (30.0, 'Call of Duty: Modern Warfare 3', 2011, 'Activision', 'X360'),
 (32.0, 'Call of Duty: Black Ops', 2010, 'Activision', 'X360'),
 (36.0, 'Call of Duty: Black Ops II', 2012, 'Activision', 'X360'),
 (62.0, 'Call of Duty: Ghosts', 2013, 'Activision', 'X360'),
 (63.0, 'Halo: Reach', 2010, 'Microsoft Game Studios', 'X360'),
 (66.0, 'Halo 4', 2012, 'Microsoft Game Studios', 'X360'),
 (73.0, 'Minecraft', 2013, 'Microsoft Game Studios', 'X360'),
 (76.0, 'The Elder Scrolls V: Skyrim', 2011, 'Bethesda Softworks', 'X360'),
 (100.0, 'Battlefield 3', 2011, 'Electronic Arts', 'X360'),
 (135.0, 'Red Dead Redemption', 2010, 'Take-Two Interactive', 'X360'),
 (140.0, 'Kinect Sports', 2010, 'Microsoft Game Studios', 'X360'),
 (141.0, 'Gears of War 3', 2011, 'Microsoft Game Studios', 'X360'),
 (173.0, "Assassin's Creed III", 2012, 'Ubisoft', 'X360'),
 (186.0, 'FIFA Soccer 13', 2012, 'Electronic Arts', 'X360'),
 (193.0, 'Fable III', 2010, 'Microsoft Game Studios', 'X360'),
 (223.0, 'Batman: Arkham City', 2011, 'Warner Bros. Interactive Entertainment', 'X360'),
 (229.0, 'Forza Motorsport 4', 2011, 'Microsoft Game Studios', 'X360'),
 (257.0, 'FIFA 14', 2013, 'Electronic Arts', 'X360')]

### Exercise 1

Figure out how to get the 20 best-selling games published by Nintendo whose genre is 'Platform'.

In [6]:
%%sql
SELECT rank, name FROM vgsale
WHERE publisher = 'Nintendo' AND genre='Platform'
LIMIT 20

 * postgresql://apicard:***@localhost/video_games
20 rows affected.


[(2.0, 'Super Mario Bros.'),
 (7.0, 'New Super Mario Bros.'),
 (9.0, 'New Super Mario Bros. Wii'),
 (19.0, 'Super Mario World'),
 (22.0, 'Super Mario Land'),
 (23.0, 'Super Mario Bros. 3'),
 (47.0, 'Super Mario 64'),
 (49.0, 'Super Mario Galaxy'),
 (54.0, 'Super Mario 3D Land'),
 (58.0, 'Super Mario All-Stars'),
 (60.0, 'Super Mario 64'),
 (65.0, 'New Super Mario Bros. 2'),
 (72.0, 'Donkey Kong Country'),
 (92.0, 'Super Mario Galaxy 2'),
 (97.0, 'Super Mario Bros. 2'),
 (126.0, 'Donkey Kong Country Returns'),
 (136.0, 'Super Mario Sunshine'),
 (163.0, 'Super Mario Advance'),
 (165.0, 'Super Mario World'),
 (176.0, 'Donkey Kong 64')]

## Sorting

So far, we've only looked at highest-ranked games and, since the table lists them first, we've been able to look at the top 10 or 20 just by setting a `LIMIT`. What if the data aren't in the order that we want to see them in? We use the `ORDER BY` command to sort them by some other criteria. 

For example, to see the games in the order in which they were published, we could run:

In [7]:
%%sql

SELECT name, year, publisher, global_sales
FROM vgsale 
ORDER BY year 
LIMIT 10

 * postgresql://apicard:***@localhost/video_games
10 rows affected.


[('Ice Hockey', 1980, 'Activision', 0.49),
 ('Checkers', 1980, 'Atari', 0.24),
 ('Kaboom!', 1980, 'Activision', 1.15),
 ('Boxing', 1980, 'Activision', 0.77),
 ('Freeway', 1980, 'Activision', 0.34),
 ('Bridge', 1980, 'Activision', 0.27),
 ('Asteroids', 1980, 'Atari', 4.31),
 ('Missile Command', 1980, 'Atari', 2.76),
 ('Defender', 1980, 'Atari', 1.05),
 ('Ms. Pac-Man', 1981, 'Atari', 1.65)]

Of course, we can mix `WHERE` and `ORDER BY`, to get only the more recent games in the order they were published. 

In [8]:
%%sql
SELECT name, year, publisher, global_sales
FROM vgsale 
WHERE year > 2015
ORDER BY year 
LIMIT 10

 * postgresql://apicard:***@localhost/video_games
10 rows affected.


[("Uncharted 4: A Thief's End", 2016, 'Sony Computer Entertainment', 4.2),
 ("Tom Clancy's The Division", 2016, 'Ubisoft', 3.61),
 ('Far Cry: Primal', 2016, 'Ubisoft', 2.13),
 ("Tom Clancy's The Division", 2016, 'Ubisoft', 2.01),
 ('Overwatch', 2016, 'Activision', 1.73),
 ("No Man's Sky", 2016, 'Hello Games', 1.6),
 ('Dark Souls III', 2016, 'Namco Bandai Games', 1.56),
 ('FIFA 17', 2016, 'Electronic Arts', 1.53),
 ('Doom (2016)', 2016, 'Bethesda Softworks', 1.39),
 ('FIFA 17', 2016, 'Electronic Arts', 4.77)]

### Exercise 2

Get the names and platforms of all the games published in 2014 in alphabetical order of their names (note that computers sort numbers before letters so you'll get some games whose name starts with some numbers before the first 'A' game).


In [9]:
%%sql

SELECT name, platform
FROM vgsale
WHERE year = 2014
ORDER BY name
LIMIT 20


 * postgresql://apicard:***@localhost/video_games
20 rows affected.


[('12-Sai. Honto no Kimochi', '3DS'),
 ('2014 FIFA World Cup Brazil', 'PS3'),
 ('2014 FIFA World Cup Brazil', 'X360'),
 ('3rd Super Robot Wars Z Jigoku Hen', 'PSV'),
 ('3rd Super Robot Wars Z Jigoku Hen', 'PS3'),
 ('A-Train: City Simulator', '3DS'),
 ('AMNESIA World', 'PSV'),
 ('Adventure Time: The Secret of the Nameless', '3DS'),
 ('Adventure Time: The Secret of the Nameless', 'X360'),
 ('Adventure Time: The Secret of the Nameless Kingdom', '3DS'),
 ('Adventure Time: The Secret of the Nameless Kingdom', 'X360'),
 ('Aikatsu! 365 Idol Days', '3DS'),
 ("Akiba's Trip: Undead & Undressed", 'PS4'),
 ('Alien: Isolation', 'PC'),
 ('Alien: Isolation', 'PS4'),
 ('Alien: Isolation', 'XOne'),
 ('Alien: Isolation', 'X360'),
 ('Alien: Isolation', 'PS3'),
 ('Amagami', 'PSV'),
 ('Amnesia Later', 'PSV')]

## Aggregation

One very powerful feature of SQL is that it allows us to create summary information by grouping rows together.
For example, we could ask ourselves which publishers have published games that have sold more than 10 million copies and how many such games did they make?

In [10]:
%%sql
SELECT publisher, COUNT(*) 
FROM vgsale 
WHERE global_sales > 10 
GROUP BY publisher 

 * postgresql://apicard:***@localhost/video_games
6 rows affected.


[('Activision', 10),
 ('Microsoft Game Studios', 2),
 ('Nintendo', 37),
 ('Sony Computer Entertainment', 4),
 ('Take-Two Interactive', 8),
 ('Ubisoft', 1)]

`GROUP BY publisher` takes all the rows with a given publisher and produces a single row in the result. This means that we need to tell SQL how we want to combine the other columns' values into a single row. The above example uses `COUNT (*)` which reports of the number of rows that were combined.

Let's take a closer look at the 4 rows for Sony:

In [11]:
%%sql
SELECT * 
FROM vgsale 
WHERE global_sales > 10 
AND publisher = 'Sony Computer Entertainment'

 * postgresql://apicard:***@localhost/video_games
4 rows affected.


[(29.0, 'Gran Turismo 3: A-Spec', 'PS2', 2001, 'Racing', 'Sony Computer Entertainment', 6.85, 5.09, 1.87, 1.16, 14.98),
 (48.0, 'Gran Turismo 4', 'PS2', 2004, 'Racing', 'Sony Computer Entertainment', 3.01, 0.01, 1.1, 7.53, 11.66),
 (53.0, 'Gran Turismo', 'PS', 1997, 'Racing', 'Sony Computer Entertainment', 4.02, 3.87, 2.54, 0.52, 10.95),
 (55.0, 'Gran Turismo 5', 'PS3', 2010, 'Racing', 'Sony Computer Entertainment', 2.96, 4.88, 0.81, 2.12, 10.77)]

Aggregating the values for `publisher` is not hard, since they're all the same, SQL just gives us a single copy of the publisher name. Other columns, we need to either ignore (causing them to be omitted from the output) or specify a way to aggregate them. 

If we don't specify a way to aggregate the value, SQL will complain. For example, the following query should fail:

In [12]:
%%sql
SELECT publisher, genre
FROM vgsale
WHERE global_sales > 10 
GROUP BY publisher 

 * postgresql://apicard:***@localhost/video_games
(psycopg2.ProgrammingError) column "vgsale.genre" must appear in the GROUP BY clause or be used in an aggregate function
LINE 1: SELECT publisher, genre
                          ^
 [SQL: 'SELECT publisher, genre\nFROM vgsale\nWHERE global_sales > 10 \nGROUP BY publisher']


Even though genre is all 'Racing' for Sony, SQL doesn't know how to combine the values and errors out. 

We must specify an aggregate function for any column that we `SELECT` in our query (except the column that we're grouping by) in order for the command to succeed.

There are [many such functions](https://www.postgresql.org/docs/9.5/functions-aggregate.html). Some common ones include:

- `SUM`: To add the values together
- `AVG`: To compute the mean of the values
- `MIN` or `MAX`: To compute the minimum and maximum respectively
    
So we could for example compute the total number of copies of these top-selling games that were sold by running:

In [13]:
%%sql
SELECT publisher, SUM(global_sales)
FROM vgsale
WHERE global_sales > 10 
GROUP BY publisher 

 * postgresql://apicard:***@localhost/video_games
6 rows affected.


[('Activision', 132.0),
 ('Microsoft Game Studios', 33.96),
 ('Nintendo', 793.05),
 ('Sony Computer Entertainment', 48.36),
 ('Take-Two Interactive', 121.41),
 ('Ubisoft', 10.26)]

Note that this does not mean that Nintendo has sold a total of 793 million games in total. 
This means that Nintendo has sold 793 million copies of its games that sold more than 10 million copies. 

The `WHERE global_sales > 10` portion of the query removes all games that sold 10 million copies or less BEFORE
the grouping happens so those games are excluded from the result. 

If we wish to filter the results AFTER the grouping
happens, we need to use the `HAVING` command. For example, to see all publishers that have sold a total of 50 million games or more we would run:

In [14]:
%%sql
SELECT publisher, SUM(global_sales)
FROM vgsale
GROUP BY publisher
HAVING SUM(global_sales) >= 50 

 * postgresql://apicard:***@localhost/video_games
25 rows affected.


[('Namco Bandai Games', 254.09),
 ('Warner Bros. Interactive Entertainment', 153.89),
 ('Eidos Interactive', 98.98),
 ('Bethesda Softworks', 82.14),
 ('LucasArts', 87.34),
 ('Microsoft Game Studios', 245.79),
 ('Electronic Arts', 1110.32),
 ('Nintendo', 1786.56),
 ('Take-Two Interactive', 399.54),
 ('Activision', 727.461),
 ('Midway Games', 69.8499),
 ('Sony Computer Entertainment', 607.502),
 ('Konami Digital Entertainment', 283.639),
 ('SquareSoft', 57.65),
 ('Ubisoft', 474.72),
 ('Atari', 157.22),
 ('505 Games', 55.91),
 ('Tecmo Koei', 53.55),
 ('THQ', 340.77),
 ('Acclaim Entertainment', 64.14),
 ('Square Enix', 145.18),
 ('Sega', 272.989),
 ('Disney Interactive Studios', 119.96),
 ('Capcom', 200.89),
 ('Vivendi Games', 58.21)]

Nintendo sales are now totalling 1786 million. That means that nearly 
1 billion of their sales came from titles that didn't sell 10 million copies (1786 million total - 793 million from the previous query).

### Exercise 3

Compute the average number of games sold by Nintendo each year (of publication) and list them in chronological order.

In [15]:
%%sql
SELECT year, AVG(global_sales)
FROM vgsale
WHERE publisher = 'Nintendo'
GROUP BY year
ORDER BY year

 * postgresql://apicard:***@localhost/video_games
35 rows affected.


[(1983, 1.82666669289271),
 (1984, 5.0622221827507),
 (1985, 7.13571453945977),
 (1986, 2.6966667274634),
 (1987, 2.98750007152557),
 (1988, 6.07333346207937),
 (1989, 7.09777775075701),
 (1990, 5.07000006948199),
 (1991, 1.22846154868603),
 (1992, 3.81100005954504),
 (1993, 2.22666669223044),
 (1994, 3.12375002354383),
 (1995, 1.67199999392033),
 (1996, 4.33529419057509),
 (1997, 1.98461541877343),
 (1998, 2.84764709367472),
 (1999, 3.26650002449751),
 (2000, 1.48043477016947),
 (2001, 2.06227271292697),
 (2002, 2.19590910456397),
 (2003, 1.41259257109077),
 (2004, 1.06403508017722),
 (2005, 2.83266667851971),
 (2006, 3.87943394896838),
 (2007, 2.48047620355196),
 (2008, 2.85062498191837),
 (2009, 4.0278125226032),
 (2010, 2.18107140150719),
 (2011, 1.98192307059295),
 (2012, 1.82161289537626),
 (2013, 2.29521740713845),
 (2014, 2.43249999955297),
 (2015, 0.846250001748558),
 (2016, 0.347000000439584),
 (None, 0.304285710943597)]

### Exercise 3 (continued)

List the years in which Sports game (genre) have sold more than 60 million copies and the number of those games sold that year (Hint: you'll need to use both `WHERE` and `HAVING`)

In [16]:
%%sql
SELECT year, SUM(global_sales)
FROM vgsale
WHERE genre='Sports'
GROUP BY year
HAVING SUM(global_sales) > 60
ORDER BY year

 * postgresql://apicard:***@localhost/video_games
7 rows affected.


[(2002, 65.42),
 (2004, 63.68),
 (2006, 136.16),
 (2007, 98.1999),
 (2008, 95.3399),
 (2009, 138.52),
 (2010, 92.53)]

## Joining

It is frequently the case that the data we need is spread across multiple tables in our database. For example, we might want to store information about the number of gaming units of each platform that have been sold and we could store that in a table called `platformsale`.


In [17]:
%%sql
select * from platformsale limit 10

 * postgresql://apicard:***@localhost/video_games
10 rows affected.


[('PlayStation 2', 'Sony', 2000, 155.0, 'PS2'),
 ('Nintendo DS', 'Nintendo', 2004, 154.02, 'DS'),
 ('Game Boy', 'Nintendo', 1989, 118.69, 'GB'),
 ('PlayStation', 'Sony', 1994, 102.49, 'PS'),
 ('Wii', 'Nintendo', 2006, 101.63, 'Wii'),
 ('PlayStation 4', 'Sony', 2013, 86.1, 'PS4'),
 ('Xbox 360', 'Microsoft', 2005, 84.0, 'X360'),
 ('PlayStation 3', 'Sony', 2006, 83.8, 'PS3'),
 ('PlayStation Portable', 'Sony', 2004, 82.0, 'PSP'),
 ('Game Boy Advance', 'Nintendo', 2001, 81.51, 'GBA')]

This means that we now have the data to answer questions like "What percentage of Wii had Wii Sports installed?" but the data are spread across two tables?

We could imagine storing the `units_sold` column in our `vgsale` table since we list the platform for each game but there are a few important reasons why that's a bad idea:

1. We would waste space by duplicating data (not a big deal for this example but a real concern for large systems)
2. Updating data (for example new Xbox 360 sales numbers since that system is still on sale) would require updating
   each row in `vgsale` that refers to that platform. This is time-consuming and error-prone. 
   
Instead we leave the data in two separate tables and need a way to 'join' the values together. We can do that by just listing multiple table names but the result is a mess:

In [18]:
%%sql 
SELECT *
FROM vgsale, platformsale
LIMIT 10

 * postgresql://apicard:***@localhost/video_games
10 rows affected.


[(1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'PlayStation 2', 'Sony', 2000, 155.0, 'PS2'),
 (1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'Nintendo DS', 'Nintendo', 2004, 154.02, 'DS'),
 (1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'Game Boy', 'Nintendo', 1989, 118.69, 'GB'),
 (1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'PlayStation', 'Sony', 1994, 102.49, 'PS'),
 (1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'Wii', 'Nintendo', 2006, 101.63, 'Wii'),
 (1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'PlayStation 4', 'Sony', 2013, 86.1, 'PS4'),
 (1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'Xbox 360', 'Microsoft', 2005, 84.0, 'X360'),
 (1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'PlayStation 3', 'Sony', 2006, 83.8, 'PS3'),
 (1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'PlayStation Portable', 'Sony', 2004, 82.0, 'PSP'),
 (1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'Game Boy Advance', 'Nintendo', 2001, 81.51, 'GBA')]

If you look carefully you might notice that the rows are identical for the first few columns and then start to 
differ after global sales. That's because SQL joins each row in the first table with each row in the second table. 
With 16,598 rows in vgsales and 40 rows in platformsale, we end up with a table of 663,920 row. 

This rarely if ever what we want. In most cases, we want to match up some aspect of the rows in the first table with 
some aspect of the rows in the second table. In most cases, we want to match up based on some column being equal.

In our video game example, the `platform` column of `vgsale` matches up with the `platform_abbreviation` column of `platformsale`. To force this match, we filter out the ones that don't have the same value for both of these columns:

In [19]:
%%sql 
SELECT *
FROM vgsale, platformsale
WHERE vgsale.platform = platformsale.platform_abbreviation
LIMIT 10


 * postgresql://apicard:***@localhost/video_games
10 rows affected.


[(1.0, 'Wii Sports', 'Wii', 2006, 'Sports', 'Nintendo', 41.49, 29.02, 3.77, 8.46, 82.74, 'Wii', 'Nintendo', 2006, 101.63, 'Wii'),
 (2.0, 'Super Mario Bros.', 'NES', 1985, 'Platform', 'Nintendo', 29.08, 3.58, 6.81, 0.77, 40.24, 'Nintendo Entertainment System', 'Nintendo', 1983, 61.91, 'NES'),
 (3.0, 'Mario Kart Wii', 'Wii', 2008, 'Racing', 'Nintendo', 15.85, 12.88, 3.79, 3.31, 35.82, 'Wii', 'Nintendo', 2006, 101.63, 'Wii'),
 (4.0, 'Wii Sports Resort', 'Wii', 2009, 'Sports', 'Nintendo', 15.75, 11.01, 3.28, 2.96, 33.0, 'Wii', 'Nintendo', 2006, 101.63, 'Wii'),
 (5.0, 'Pokemon Red/Pokemon Blue', 'GB', 1996, 'Role-Playing', 'Nintendo', 11.27, 8.89, 10.22, 1.0, 31.37, 'Game Boy', 'Nintendo', 1989, 118.69, 'GB'),
 (6.0, 'Tetris', 'GB', 1989, 'Puzzle', 'Nintendo', 23.2, 2.26, 4.22, 0.58, 30.26, 'Game Boy', 'Nintendo', 1989, 118.69, 'GB'),
 (7.0, 'New Super Mario Bros.', 'DS', 2006, 'Platform', 'Nintendo', 11.38, 9.23, 6.5, 2.9, 30.01, 'Nintendo DS', 'Nintendo', 2004, 154.02, 'DS'),
 (8.0, 'Wii Play', 'Wii', 2006, 'Misc', 'Nintendo', 14.03, 9.2, 2.93, 2.85, 29.02, 'Wii', 'Nintendo', 2006, 101.63, 'Wii'),
 (9.0, 'New Super Mario Bros. Wii', 'Wii', 2009, 'Platform', 'Nintendo', 14.59, 7.06, 4.7, 2.26, 28.62, 'Wii', 'Nintendo', 2006, 101.63, 'Wii'),
 (10.0, 'Duck Hunt', 'NES', 1984, 'Shooter', 'Nintendo', 26.93, 0.63, 0.28, 0.47, 28.31, 'Nintendo Entertainment System', 'Nintendo', 1983, 61.91, 'NES')]

Notice that the result looks more sensical: we end up with one row from vgsale and the corresponding row from
platformsale (copied multiple times since there were only 40 rows in platform sale).

We can check the size of the resulting table by running:

In [20]:
%%sql 
SELECT COUNT(*)
FROM vgsale, platformsale
WHERE vgsale.platform = platformsale.platform_abbreviation


 * postgresql://apicard:***@localhost/video_games
1 rows affected.


[(15616,)]

Note that this is slightly smaller than the 16,000+ rows that we started with in vgsales because games whose platform was not part of the list in platformsale got removed by the `WHERE` clause.

You might also see some cases where the comma between the table names is replaced with the keyword `JOIN` and `WHERE` is replaced with `ON`. This is synonymous but sometimes preferred to make it clear that you are joining two tables and that your filters are there to specify how those tables are to be joined:

In [21]:
%%sql 
SELECT name, global_sales, platformsale.platform, units_sold 
FROM vgsale JOIN platformsale 
ON vgsale.platform = platformsale.platform_abbreviation
LIMIT 10

 * postgresql://apicard:***@localhost/video_games
10 rows affected.


[('Wii Sports', 82.74, 'Wii', 101.63),
 ('Super Mario Bros.', 40.24, 'Nintendo Entertainment System', 61.91),
 ('Mario Kart Wii', 35.82, 'Wii', 101.63),
 ('Wii Sports Resort', 33.0, 'Wii', 101.63),
 ('Pokemon Red/Pokemon Blue', 31.37, 'Game Boy', 118.69),
 ('Tetris', 30.26, 'Game Boy', 118.69),
 ('New Super Mario Bros.', 30.01, 'Nintendo DS', 154.02),
 ('Wii Play', 29.02, 'Wii', 101.63),
 ('New Super Mario Bros. Wii', 28.62, 'Wii', 101.63),
 ('Duck Hunt', 28.31, 'Nintendo Entertainment System', 61.91)]

We can now use all the SQL tools that we've learned on this combined table. For example, to find out what percentage of possible units had each game installed, we can run:

In [22]:
%%sql 
SELECT name, global_sales, platformsale.platform, units_sold, global_sales/units_sold AS sale_percentage 
FROM vgsale JOIN platformsale 
ON vgsale.platform = platformsale.platform_abbreviation
WHERE units_sold is not null
ORDER BY sale_percentage DESC
LIMIT 10

 * postgresql://apicard:***@localhost/video_games
10 rows affected.


[('Wii Sports', 82.74, 'Wii', 101.63, 0.81413),
 ('Super Mario Bros.', 40.24, 'Nintendo Entertainment System', 61.91, 0.649976),
 ('Mario Kart 8', 6.96, 'Wii U', 13.56, 0.513274),
 ('Duck Hunt', 28.31, 'Nintendo Entertainment System', 61.91, 0.457277),
 ('Super Mario World', 20.61, 'Super Nintendo Entertainment System', 49.1, 0.419756),
 ('New Super Mario Bros. U', 5.19, 'Wii U', 13.56, 0.382743),
 ('Super Smash Bros. for Wii U and 3DS', 5.02, 'Wii U', 13.56, 0.370206),
 ('Super Mario 64', 11.89, 'Nintendo 64', 32.93, 0.361069),
 ('Halo 2', 8.49, 'Xbox', 24.0, 0.35375),
 ('Mario Kart Wii', 35.82, 'Wii', 101.63, 0.352455)]

From this, we conclude that 81% of Wiis had a copy of Wii Sports and 42% of Super Nintendos had Super Mario World.

### Exercise 4

Use `JOIN` to find the names of the games that were released in the same year as their corresponding platform and that sold at least 10 million copies.

In [23]:
%%sql 
SELECT rank, name
FROM vgsale JOIN platformsale 
ON vgsale.platform = platformsale.platform_abbreviation
WHERE year = released_year
AND global_sales >= 10


 * postgresql://apicard:***@localhost/video_games
9 rows affected.


[(1.0, 'Wii Sports'),
 (6.0, 'Tetris'),
 (8.0, 'Wii Play'),
 (19.0, 'Super Mario World'),
 (22.0, 'Super Mario Land'),
 (43.0, 'Mario Kart 7'),
 (47.0, 'Super Mario 64'),
 (54.0, 'Super Mario 3D Land'),
 (60.0, 'Super Mario 64')]